src/periodic.jl

```diff
@@ -0,0 +1,85 @@
+const PeriodicFunction = Union{typeof(cos), typeof(sin)}
+const PeriodicParameter = Union{Real,Period}
+
+"""
+    Periodic(f, period, phase_shift) <: Transform
+
+Applies a periodic function `f` with provided `period` and `phase_shift` to the data.
+
+# Fields
+* `f::PeriodicFunction`: the periodic function
+* `period<:PeriodicParameter`: the duration it takes for the periodic function to repeat.
+    The sign of the value determines the direction of the function. Must be non-zero.
+* `phase_shift<:PeriodicParameter`: adjusts the phase of the periodic function, measured
+    in the same units as the input. Increasing the value translates the function to the
+    right, toward higher/later input values.
+"""
+struct Periodic{T<:PeriodicParameter} <: Transform
+    f::PeriodicFunction
+    period::T
+    phase_shift::T
+end
+
+"""
+    Periodic(f::PeriodicFunction, period<:PeriodicParameter) -> Periodic
+
+A constructor for [`Periodic`](@ref).
+Returns a `Periodic` transform with zero phase shift.
+"""
+function Periodic(f, period)
+    return Periodic{typeof(period)}(f, period, zero(typeof(period)))
+end
+
+function _apply!(x::AbstractArray{T}, P::Periodic; kwargs...) where T <: Real
+    x[:] = P.f.(2π .* (x .- P.phase_shift) / P.period)
+    return x
+end
+
+function apply(x::AbstractArray{T}, P::Periodic{U}; kwargs...) where {T <: TimeType, U <: Period}
```

Comment:
I guess you overloaded `apply` because this can't be mutating since it's consuming a time-type put returning a real-type?

Reply:
Correct. And the `U <: Period` was needed to avoid method ambiguity with the `apply(table...` method. I was surprised at that because I thought `::AbstractArray{T}` would be enough to resolve that ambiguity.